### DISCUSSION TITLE
Add Image<T>.LoadPixelData variant that allows reusing an image object

### DISCUSSION THREAD

**Discussion by fredlllll** [SixLabors/ImageSharp#1472] [Ideas] [answered]
for my project i used system.drawing for prototyping, but now it has to run on linux and i wanted to use imagesharp.

in my implementation i pooled bitmaps to prevent memory shooting up and take load of the GC. i reused bitmap instances of the fitting size and set their pixels using the LockBits[etc...] method to then save the image to disc and then return the bitmap into the pool

when changing to imagesharp i didnt find a way to reuse an image when directly settings its pixels using LoadPixelData.
here you can see that a new image is created every time i call this method to convert my pixel data into an image https://github.com/SixLabors/ImageSharp/blob/78a584e8482b052d7a9885682299e2f37518d83d/src/ImageSharp/Image.LoadPixelData.cs#L132

doing this caused memory spikes as seen here
![image of memory graph showing spikes and gc events](https://i.imgur.com/4qZPj52.png)

it would be nice to have a variant of LoadPixelData that reuses a passed in image instance instead of creating a new one every time

i tried copying the logic from the LoadPixelData and changing it to my needs, but alas, some important bits are internal and cant be accessed by me =/

**Comment by antonfirsov** [comment, marked as answer]
@fredlllll we have an internal pooling mechanism implemented by the type `ArrayPoolMemoryAllocator`, so normally there is no benefit from bothering with manual reusing/pooling strategies, all you need is to make sure you properly **dispose** all images being created. Note that the Image's buffer is not the only one being created during processing, there are temporary buffers created+destroyed behind the scenes, so proper pooling is a must for getting this right, you cant achieve it by just reusing the image instance. And again -- pooling is provided out of the box by the library. See the following article for more details:
https://docs.sixlabors.com/articles/imagesharp/memorymanagement.html

However, based on the memory timeline you posted something *might* got wrong in your app, but without code/more info I can't advise. A lot depends on image sizes + what operations you do.

But first, I suggest to rethink your code based on the article above.

**Reply by fredlllll** [reply]
i just tried making a minimal example of the behaviour and for the love of it i cant recreate it. i also moved the code i used for image saving into a seperate class in my project and it stopped the behaviour. but it is still occuring in the commit 72e91928d6385305bf8fde2807c4e00c8797d2b0 in my repo https://github.com/fredlllll/MandelbrotCSharp in case you want to check that out. i must be hitting an edge case.

commenting out the using clause and moving the arraypool.Put outside so its still called removes any memory spikes https://github.com/fredlllll/MandelbrotCSharp/blob/72e91928d6385305bf8fde2807c4e00c8797d2b0/MandelbrotCSharp/MandelField.cs#L229

anyway my problem is solved, but if you want to investigate that edge case (or prove that im mistaken) id be really intersted in what caused this

**Comment by DaZombieKiller** [comment]
I've run into a similar form of this while experimenting with using ImageSharp in a Unity project. The idea was to use ImageSharp to load external images, as Unity just exposes a blocking main-thread-only `LoadImage(byte[])` method with support for PNG and JPG only. Consider the following example:

```CSharp
static Texture2D LoadImage<T>(Configuration configuration, Stream input)
    where T : unmanaged, IPixel<T>
{
    // Identify the image without loading it so we can create the Texture2D with an appropriate size.
    IImageInfo info = Image.Identify(configuration, input);
    var texture     = new Texture2D(info.Width, info.Height, GetTextureFormat<T>(), mipChain: false);
    
    // Unity allows access to the pixel data using GetPixelData<T>, which can
    // then be wrapped into a Memory<T> instance using a custom MemoryManager.
    NativeArray<T> pixels = texture.GetPixelData<T>(mipLevel: 0);
    
    // There doesn't appear to be a way to load the pixels into the image without needing to
    // allocate a new one and then copying the pixels across, which is a slow operation (omitted here).
    var image = Image.WrapMemory(configuration, pixels.AsMemory(), info.Width, info.Height, info.Metadata);

    return texture;
}
```

As seen above, I have no control over the allocation of the memory for the pixel data and am forced to pay the price of a copy via `Image.Load`. Is there perhaps an API that I've missed somewhere for loading pixel data directly into a `Memory<T>` instance?

**Reply by DaZombieKiller** [reply]
I think this situation might be best served by an `IImageMemoryDecoder` interface (ideally with a better name), exposing an API for reading pixel data directly into a `Memory<TPixel>`:

```CSharp
interface IImageMemoryDecoder
{
    /// <summary>Decodes the image from the specified stream into a contiguous buffer of a specific pixel type.</summary>
    void Decode<TPixel>(Configuration configuration, Stream stream, Memory<TPixel> memory)
        where TPixel : unmanaged, IPixel<TPixel>;

    /// <summary>Decodes the image from the specified stream into a contiguous buffer of a specific pixel type.</summary>
    Task DecodeAsync<TPixel>(Configuration configuration, Stream stream, Memory<TPixel> memory, CancellationToken cancellationToken)
        where TPixel : unmanaged, IPixel<TPixel>;
}
```

Which would allow for a fast path, should a decoder implement it:

```CSharp
static async Task<Texture2D> LoadImageAsync<T>(Configuration configuration, Stream stream, CancellationToken cancellationToken = default)
    where T : unmanaged, IPixel<T>
{
    var startPosition  = stream.Position;
    var (info, format) = await Image.IdentifyWithFormatAsync(configuration, stream, cancellationToken);
    stream.Position    = startPosition;

    // The decoder would need to be discovered manually, but that should be acceptable given
    // this is an advanced scenario that most likely doesn't need a convenience Image.* method.
    var decoder = configuration.ImageFormatsManager.FindDecoder(format);
    var texture = new Texture2D(info.Width, info.Height, GetTextureFormat<T>(), mipChain: false);
    var memory  = texture.GetPixelData<T>(mipLevel: 0).AsMemory();

    if (decoder is IImageMemoryDecoder memoryDecoder)
    {
        // Fast path: the decoder supports decoding directly into a Memory<TPixel>
        await memoryDecoder.DecodeAsync(configuration, stream, memory, cancellationToken);
    }
    else
    {
        // Slow path: we need to allocate a new image and copy it to a wrapper over our buffer
        using var image   = await decoder.DecodeAsync<T>(configuration, stream, cancellationToken);
        using var wrapper = Image.WrapMemory(configuration, memory, image.Width, image.Height, info.Metadata);
        CopyPixelData(image, wrapper);
    }

    texture.Apply();
    return texture;
}
```

This would avoid breaking changes (eg, changing the `IImageDecoder` interface), but still leaves a few questions on the table:
* How would this work with an image containing multiple frames? Decoding just the first?
* Would it make sense for the API to expose partial decode overloads for dealing with large images?
* Is it feasible to change existing implementations of `IImageDecoder` to use `IImageMemoryDecoder` internally (to reduce code duplication)?

Does this seem like something that could realistically make it into ImageSharp?

**Reply by antonfirsov** [reply]
This idea makes sense for me, I would consider opening a tracking issue *"Decode into existing `Memory<T>`"* to discuss and elaborate the details. I don't think we are able to handle it with high priority at the moment though, but I would be happy to accept community contributions for the feature. @JimBobSquarePants thoughts?

When it comes to API shape, I would prefer a solution that works with existing decoder infrastructure, and exposed through  trivially discoverable methods, hiding the `if (decoder is IImageMemoryDecoder memoryDecoder) ... else ...` (and similar) stuff from the user:
- Either by having static overloads `Image.Load[Async](...., Memory<T>)`
- Or an instance method `image.Load(stream)` which can be used in combination with `WrapMemory`.

The `IImageMemoryDecoder` or whatever underlying interface should be designed in a way that supports the main API the best.

@DaZombieKiller are you interested in PR-ing this stuff?

**Reply by DaZombieKiller** [reply]
@antonfirsov I'd certainly be happy to take a shot at PR-ing this at some point. I've opened a new discussion for the feature here: https://github.com/SixLabors/ImageSharp/discussions/1487

**Reply by JimBobSquarePants** [reply]
@antonfirsov I don't like it. We already have the new Image.WrapMemory overloads via https://github.com/SixLabors/ImageSharp/pull/1419 available in the nightlies which performs a copy free wrap.

**Reply by DaZombieKiller** [reply]
`WrapMemory` works when you want to use an existing buffer of decoded pixels as an `Image`, but there's currently no way to use ImageSharp to decode directly into an existing buffer or `Image`, thus necessitating copies.

**Reply by JimBobSquarePants** [reply]
We're talking in two different places now. You should be petitioning Unity to add means to create textures from buffers without copying.
